File: go/aletheia/ltl.go
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause

package aletheia

// Predicate is a test on one signal's value, the leaf of a formula. Only the
// types below implement it: the method it requires is unexported, so a value
// of this interface is always one the kernel knows.
//
// Every value a predicate carries is an exact [Rational], never a float, which
// is what lets the kernel decide the comparison rather than approximate it.
// Build one with [IntRational] for a whole number, or with [FromDecimal] for
// decimal text, which the kernel parses.
type Predicate interface {
	predicate() // sealed
}

// Equals tests whether a signal's value equals a specific value.
type Equals struct {
	Signal SignalName
	Value  Rational
}

// LessThan tests whether a signal's value is strictly less than a threshold.
type LessThan struct {
	Signal SignalName
	Value  Rational
}

// GreaterThan tests whether a signal's value is strictly greater than a threshold.
type GreaterThan struct {
	Signal SignalName
	Value  Rational
}

// LessThanOrEqual tests whether a signal's value is at most a threshold.
type LessThanOrEqual struct {
	Signal SignalName
	Value  Rational
}

// GreaterThanOrEqual tests whether a signal's value is at least a threshold.
type GreaterThanOrEqual struct {
	Signal SignalName
	Value  Rational
}

// Between tests whether a signal's value is within [Min, Max].
type Between struct {
	Signal SignalName
	Min    Rational
	Max    Rational
}

// ChangedBy tests how far a signal moved since the previous frame. A positive
// delta asks for a rise of at least that much, a negative one for a fall.
type ChangedBy struct {
	Signal SignalName
	Delta  Rational
}

// StableWithin tests whether a signal's value stayed within Tolerance of its previous value.
type StableWithin struct {
	Signal    SignalName
	Tolerance Rational
}

func (Equals) predicate()             {}
func (LessThan) predicate()           {}
func (GreaterThan) predicate()        {}
func (LessThanOrEqual) predicate()    {}
func (GreaterThanOrEqual) predicate() {}
func (Between) predicate()            {}
func (ChangedBy) predicate()          {}
func (StableWithin) predicate()       {}

// SignalBuilder names a signal and builds every predicate the kernel has over
// it: the five comparisons, the range, and the two that read the previous
// frame as well as this one.
//
// This is the formula-level builder. [CheckSignal] is the other one, which
// names and builds whole checks.
type SignalBuilder struct {
	name SignalName
}

// Signal names the signal a predicate is about:
//
//	aletheia.Signal("Speed").LessThanOrEqual(aletheia.IntRational(220))
//
// The predicate it returns goes inside [Atomic], or straight into [Always],
// [Eventually] and the rest through it.
func Signal(name string) SignalBuilder { return SignalBuilder{name: SignalName(name)} }

// None of these can fail. A threshold is already an exact rational, and the
// two that can be asked for something impossible, a range whose minimum is
// above its maximum and a negative tolerance, are refused where the property
// is serialised, which is the one place every route to a predicate passes
// through: a caller may also write the value directly.

// Equals holds when the signal's value is exactly v.
func (s SignalBuilder) Equals(v Rational) Predicate { return Equals{Signal: s.name, Value: v} }

// LessThan holds when the value is below v.
func (s SignalBuilder) LessThan(v Rational) Predicate { return LessThan{Signal: s.name, Value: v} }

// GreaterThan holds when the value is above v.
func (s SignalBuilder) GreaterThan(v Rational) Predicate {
	return GreaterThan{Signal: s.name, Value: v}
}

// LessThanOrEqual holds when the value is at most v.
func (s SignalBuilder) LessThanOrEqual(v Rational) Predicate {
	return LessThanOrEqual{Signal: s.name, Value: v}
}

// GreaterThanOrEqual holds when the value is at least v.
func (s SignalBuilder) GreaterThanOrEqual(v Rational) Predicate {
	return GreaterThanOrEqual{Signal: s.name, Value: v}
}

// Between holds when the value is at least lo and at most hi.
func (s SignalBuilder) Between(lo, hi Rational) Predicate {
	return Between{Signal: s.name, Min: lo, Max: hi}
}

// ChangedBy holds when the signal moved by at least delta since the previous
// frame, a positive delta asking for a rise and a negative one for a fall.
func (s SignalBuilder) ChangedBy(delta Rational) Predicate {
	return ChangedBy{Signal: s.name, Delta: delta}
}

// StableWithin holds when the signal stayed within tolerance of its value in
// the previous frame.
func (s SignalBuilder) StableWithin(tolerance Rational) Predicate {
	return StableWithin{Signal: s.name, Tolerance: tolerance}
}

// Formula is a property over a trace of frames, built from predicates. Only
// the types below implement it, the method it requires being unexported.
type Formula interface {
	formula() // sealed
}

// Atomic wraps a Predicate as a formula leaf.
type Atomic struct{ Predicate Predicate }

// Not is the logical negation of a formula.
type Not struct{ Inner Formula }

// And is the logical conjunction of two formulas.
type And struct{ Left, Right Formula }

// Or is the logical disjunction of two formulas.
type Or struct{ Left, Right Formula }

// Next holds if the inner formula holds at the next time step.
type Next struct{ Inner Formula }

// WeakNext holds if the inner formula holds at the next time step, or
// vacuously at the end of the trace (no successor frame).
type WeakNext struct{ Inner Formula }

// Always holds if the inner formula holds at every future time step.
type Always struct{ Inner Formula }

// Eventually holds if the inner formula holds at some future time step.
type Eventually struct{ Inner Formula }

// Until holds if Left holds until Right becomes true.
type Until struct{ Left, Right Formula }

// Release holds if Right holds until (and including when) Left becomes true.
type Release struct{ Left, Right Formula }

// MetricAlways holds if the inner formula holds for every step within Bound.
type MetricAlways struct {
	Bound TimeBound
	Inner Formula
}

// MetricEventually holds if the inner formula holds at some step within Bound.
type MetricEventually struct {
	Bound TimeBound
	Inner Formula
}

// MetricUntil is a time-bounded Until: Left holds until Right within Bound.
type MetricUntil struct {
	Bound TimeBound
	Left  Formula
	Right Formula
}

// MetricRelease is a time-bounded Release: Right holds until Left within Bound.
type MetricRelease struct {
	Bound TimeBound
	Left  Formula
	Right Formula
}

func (Atomic) formula()           {}
func (Not) formula()              {}
func (And) formula()              {}
func (Or) formula()               {}
func (Next) formula()             {}
func (WeakNext) formula()         {}
func (Always) formula()           {}
func (Eventually) formula()       {}
func (Until) formula()            {}
func (Release) formula()          {}
func (MetricAlways) formula()     {}
func (MetricEventually) formula() {}
func (MetricUntil) formula()      {}
func (MetricRelease) formula()    {}

// The shapes common enough to have a name of their own.

// Never holds when the predicate holds at no frame.
func Never(p Predicate) Formula {
	return Always{Inner: Not{Inner: Atomic{Predicate: p}}}
}

// AlwaysWithin holds when the formula holds at every frame inside the bound.
func AlwaysWithin(bound TimeBound, f Formula) Formula {
	return MetricAlways{Bound: bound, Inner: f}
}

// EventuallyWithin holds when the formula holds at some frame inside the bound.
func EventuallyWithin(bound TimeBound, f Formula) Formula {
	return MetricEventually{Bound: bound, Inner: f}
}

// Implies holds when the consequent holds at every frame the antecedent does,
// which is the disjunction the logic has rather than a connective of its own.
func Implies(antecedent, consequent Formula) Formula {
	return Or{Left: Not{Inner: antecedent}, Right: consequent}
}

File: rust/src/ltl.rs
// SPDX-FileCopyrightText: 2025 Nicolas Pelletier
// SPDX-License-Identifier: BSD-2-Clause

//! LTL predicates and formulas.
//!
//! `Predicate` and `Formula` are plain Rust enums — naturally sealed and
//! exhaustively matchable, replacing the Go binding's `predicate()` / `formula()`
//! sealed-interface marker methods with no ceremony. `to_value` emits the exact
//! JSON wire shape the Agda core accepts (`{"operator":…}` for formulas,
//! `{"predicate":…}` for atoms), enforcing the same validity rules the other
//! bindings apply at serialization time.

use serde_json::{json, Value};

use crate::error::Error;
use crate::types::{Rational, TimeBound};

/// Client-side recursion guard on formula nesting depth (mirrors Go's 100-deep
/// stack guard), applied before serialization to fail fast rather than recurse
/// unboundedly. This is **distinct from** the kernel's JSON nesting cap of 64
/// (`Aletheia.Limits.max-nesting-depth`), which is the true wire limit: a formula
/// deeper than 64 is rejected by the core as a `nesting_depth` bound even though
/// it passes this guard.
pub const MAX_FORMULA_DEPTH: usize = 100;

/// A signal predicate — an atomic proposition over one signal's value.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Predicate {
    /// `signal == value`.
    Equals { signal: String, value: Rational },
    /// `signal < value`.
    LessThan { signal: String, value: Rational },
    /// `signal > value`.
    GreaterThan { signal: String, value: Rational },
    /// `signal <= value`.
    LessThanOrEqual { signal: String, value: Rational },
    /// `signal >= value`.
    GreaterThanOrEqual { signal: String, value: Rational },
    /// `min <= signal <= max` (requires `min <= max`).
    Between {
        signal: String,
        min: Rational,
        max: Rational,
    },
    /// The signal changed by at least `delta` (sign selects direction).
    ChangedBy { signal: String, delta: Rational },
    /// The signal stayed within `tolerance` of its previous value (`tolerance >= 0`).
    StableWithin { signal: String, tolerance: Rational },
}

/// Compare two rationals with positive denominators: `a <= b`.
fn rational_le(a: Rational, b: Rational) -> bool {
    // Denominators are guaranteed positive by `Rational::new`, so the direction
    // of the inequality is preserved by cross-multiplication. i128 avoids i64
    // overflow on the product.
    i128::from(a.numerator()) * i128::from(b.denominator())
        <= i128::from(b.numerator()) * i128::from(a.denominator())
}

impl Predicate {
    /// `signal == value` (Agda `ValuePredicate.Equals`).
    #[must_use]
    pub fn equals(signal: impl Into<String>, value: impl Into<Rational>) -> Predicate {
        Predicate::Equals {
            signal: signal.into(),
            value: value.into(),
        }
    }

    /// `signal < value` (Agda `ValuePredicate.LessThan`).
    #[must_use]
    pub fn less_than(signal: impl Into<String>, value: impl Into<Rational>) -> Predicate {
        Predicate::LessThan {
            signal: signal.into(),
            value: value.into(),
        }
    }

    /// `signal > value` (Agda `ValuePredicate.GreaterThan`).
    #[must_use]
    pub fn greater_than(signal: impl Into<String>, value: impl Into<Rational>) -> Predicate {
        Predicate::GreaterThan {
            signal: signal.into(),
            value: value.into(),
        }
    }

    /// `signal <= value` (Agda `ValuePredicate.LessThanOrEqual`).
    #[must_use]
    pub fn less_than_or_equal(signal: impl Into<String>, value: impl Into<Rational>) -> Predicate {
        Predicate::LessThanOrEqual {
            signal: signal.into(),
            value: value.into(),
        }
    }

    /// `signal >= value` (Agda `ValuePredicate.GreaterThanOrEqual`).
    #[must_use]
    pub fn greater_than_or_equal(
        signal: impl Into<String>,
        value: impl Into<Rational>,
    ) -> Predicate {
        Predicate::GreaterThanOrEqual {
            signal: signal.into(),
            value: value.into(),
        }
    }

    /// Encode the predicate as its `{"predicate":…}` wire object.
    ///
    /// # Errors
    /// [`Error::Validation`] if `Between.min > max` or `StableWithin.tolerance < 0`.
    pub(crate) fn to_value(&self) -> Result<Value, Error> {
        let v = match self {
            Predicate::Equals { signal, value } => {
                json!({ "predicate": "equals", "signal": signal, "value": value.to_value() })
            }
            Predicate::LessThan { signal, value } => {
                json!({ "predicate": "lessThan", "signal": signal, "value": value.to_value() })
            }
            Predicate::GreaterThan { signal, value } => {
                json!({ "predicate": "greaterThan", "signal": signal, "value": value.to_value() })
            }
            Predicate::LessThanOrEqual { signal, value } => {
                json!({ "predicate": "lessThanOrEqual", "signal": signal, "value": value.to_value() })
            }
            Predicate::GreaterThanOrEqual { signal, value } => {
                json!({ "predicate": "greaterThanOrEqual", "signal": signal, "value": value.to_value() })
            }
            Predicate::Between { signal, min, max } => {
                if !rational_le(*min, *max) {
                    return Err(Error::Validation(format!(
                        "between: min ({}/{}) exceeds max ({}/{})",
                        min.numerator(),
                        min.denominator(),
                        max.numerator(),
                        max.denominator()
                    )));
                }
                json!({ "predicate": "between", "signal": signal,
                        "min": min.to_value(), "max": max.to_value() })
            }
            Predicate::ChangedBy { signal, delta } => {
                json!({ "predicate": "changedBy", "signal": signal, "delta": delta.to_value() })
            }
            Predicate::StableWithin { signal, tolerance } => {
                if tolerance.numerator() < 0 {
                    return Err(Error::Validation(format!(
                        "stableWithin: negative tolerance {}/{}",
                        tolerance.numerator(),
                        tolerance.denominator()
                    )));
                }
                json!({ "predicate": "stableWithin", "signal": signal,
                        "tolerance": tolerance.to_value() })
            }
        };
        Ok(v)
    }
}

/// An LTL formula over signal predicates.
///
/// Recursive variants box their children. Construct directly, or via the
/// convenience constructors ([`Formula::never`], [`Formula::implies`], …).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Formula {
    /// A predicate leaf.
    Atomic(Predicate),
    /// Logical negation.
    Not(Box<Formula>),
    /// Conjunction.
    And(Box<Formula>, Box<Formula>),
    /// Disjunction.
    Or(Box<Formula>, Box<Formula>),
    /// Holds at the next step.
    Next(Box<Formula>),
    /// Holds at the next step, or vacuously at end-of-trace.
    WeakNext(Box<Formula>),
    /// Holds at every future step (G).
    Always(Box<Formula>),
    /// Holds at some future step (F).
    Eventually(Box<Formula>),
    /// `left` holds until `right` becomes true.
    Until(Box<Formula>, Box<Formula>),
    /// `right` holds up to and including when `left` becomes true.
    Release(Box<Formula>, Box<Formula>),
    /// `inner` holds at every step within `bound`.
    MetricAlways {
        bound: TimeBound,
        inner: Box<Formula>,
    },
    /// `inner` holds at some step within `bound`.
    MetricEventually {
        bound: TimeBound,
        inner: Box<Formula>,
    },
    /// Time-bounded until within `bound`.
    MetricUntil {
        bound: TimeBound,
        left: Box<Formula>,
        right: Box<Formula>,
    },
    /// Time-bounded release within `bound`.
    MetricRelease {
        bound: TimeBound,
        left: Box<Formula>,
        right: Box<Formula>,
    },
}

impl Formula {
    /// `G(¬p)` — the predicate never holds.
    #[must_use]
    pub fn never(p: Predicate) -> Formula {
        Formula::Always(Box::new(Formula::Not(Box::new(Formula::Atomic(p)))))
    }

    /// `¬antecedent ∨ consequent` — the standard LTL encoding of implication.
    #[must_use]
    pub fn implies(antecedent: Formula, consequent: Formula) -> Formula {
        Formula::Or(
            Box::new(Formula::Not(Box::new(antecedent))),
            Box::new(consequent),
        )
    }

    /// `G[0,bound] inner`.
    #[must_use]
    pub fn always_within(bound: TimeBound, inner: Formula) -> Formula {
        Formula::MetricAlways {
            bound,
            inner: Box::new(inner),
        }
    }

    /// `F[0,bound] inner`.
    #[must_use]
    pub fn eventually_within(bound: TimeBound, inner: Formula) -> Formula {
        Formula::MetricEventually {
            bound,
            inner: Box::new(inner),
        }
    }

    /// Encode the formula as its `{"operator":…}` wire object.
    ///
    /// # Errors
    /// [`Error::Validation`] if nesting exceeds [`MAX_FORMULA_DEPTH`] or a
    /// contained predicate is invalid.
    pub(crate) fn to_value(&self) -> Result<Value, Error> {
        self.to_value_depth(0)
    }

    fn to_value_depth(&self, depth: usize) -> Result<Value, Error> {
        if depth > MAX_FORMULA_DEPTH {
            return Err(Error::Validation(format!(
                "formula nesting depth exceeds {MAX_FORMULA_DEPTH}"
            )));
        }
        let unary = |op: &str, inner: &Formula| -> Result<Value, Error> {
            Ok(json!({ "operator": op, "formula": inner.to_value_depth(depth + 1)? }))
        };
        let binary = |op: &str, l: &Formula, r: &Formula| -> Result<Value, Error> {
            Ok(json!({ "operator": op,
                       "left": l.to_value_depth(depth + 1)?,
                       "right": r.to_value_depth(depth + 1)? }))
        };
        let metric_unary = |op: &str, b: TimeBound, inner: &Formula| -> Result<Value, Error> {
            Ok(json!({ "operator": op, "timebound": b.micros(),
                       "formula": inner.to_value_depth(depth + 1)? }))
        };
        let metric_binary =
            |op: &str, b: TimeBound, l: &Formula, r: &Formula| -> Result<Value, Error> {
                Ok(json!({ "operator": op, "timebound": b.micros(),
                           "left": l.to_value_depth(depth + 1)?,
                           "right": r.to_value_depth(depth + 1)? }))
            };
        match self {
            Formula::Atomic(p) => Ok(json!({ "operator": "atomic", "predicate": p.to_value()? })),
            Formula::Not(f) => unary("not", f),
            Formula::And(l, r) => binary("and", l, r),
            Formula::Or(l, r) => binary("or", l, r),
            Formula::Next(f) => unary("next", f),
            Formula::WeakNext(f) => unary("weakNext", f),
            Formula::Always(f) => unary("always", f),
            Formula::Eventually(f) => unary("eventually", f),
            Formula::Until(l, r) => binary("until", l, r),
            Formula::Release(l, r) => binary("release", l, r),
            Formula::MetricAlways { bound, inner } => metric_unary("metricAlways", *bound, inner),
            Formula::MetricEventually { bound, inner } => {
                metric_unary("metricEventually", *bound, inner)
            }
            Formula::MetricUntil { bound, left, right } => {
                metric_binary("metricUntil", *bound, left, right)
            }
            Formula::MetricRelease { bound, left, right } => {
                metric_binary("metricRelease", *bound, left, right)
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Predicate;
    use crate::types::Rational;

    #[test]
    fn equals_constructor_builds_equals_variant() {
        assert_eq!(
            Predicate::equals("Speed", 220),
            Predicate::Equals {
                signal: "Speed".to_string(),
                value: Rational::integer(220),
            }
        );
    }

    #[test]
    fn less_than_constructor_builds_less_than_variant() {
        assert_eq!(
            Predicate::less_than("Speed", 220),
            Predicate::LessThan {
                signal: "Speed".to_string(),
                value: Rational::integer(220),
            }
        );
    }

    #[test]
    fn greater_than_constructor_builds_greater_than_variant() {
        assert_eq!(
            Predicate::greater_than("Speed", 220),
            Predicate::GreaterThan {
                signal: "Speed".to_string(),
                value: Rational::integer(220),
            }
        );
    }

    #[test]
    fn less_than_or_equal_constructor_builds_lte_variant() {
        assert_eq!(
            Predicate::less_than_or_equal("Speed", 220),
            Predicate::LessThanOrEqual {
                signal: "Speed".to_string(),
                value: Rational::integer(220),
            }
        );
    }

    #[test]
    fn greater_than_or_equal_constructor_builds_gte_variant() {
        assert_eq!(
            Predicate::greater_than_or_equal("Speed", 220),
            Predicate::GreaterThanOrEqual {
                signal: "Speed".to_string(),
                value: Rational::integer(220),
            }
        );
    }

    #[test]
    fn fractional_value_via_rational() {
        // A fractional threshold is passed as an explicit Rational (no From<f64>).
        let half = Rational::new(1, 2).expect("1/2 is a valid rational");
        assert_eq!(
            Predicate::less_than("Throttle", half),
            Predicate::LessThan {
                signal: "Throttle".to_string(),
                value: half,
            }
        );
    }
}
